[WIP] Bots (testing): Cleaner check_expected_responses implementation

neiljp opened the pull request:
I'm aware that @derAnfaenger indicated that the test framework was now 'stable', but I've been working on this a little recently after mentioning to @showell how the approach of all the extra keyword arguments just felt a little wrong, so thought I'd better get it out there.

While some of the changes to that interface are my doing, I was trying to write tests for pollbot and realised the interface was not exactly obvious; in trying to improve that, I've arrived at a solution which:
* I believe is as simple as the previous case for straight-forward cases (good defaults)
* Works so as to extend further the input/output in an 'expected' specification
* Allows specification of defaults (method, response_template (for send_message) and results)
* Allows (almost fully!) specialising each expected case to use a different method/response[/result]
* Defaults to the two default message input types (private/stream) now specified in the class, but can be generalised by naming one dict entry (or specialising the default as in followup_response)
* Supports testing the multiple-response case (message -> send_*  x n)
* Supports code testing success/fail of the send_* functions

I've adapted test_followup.py as an example, but have a local test_poll.py which also makes use of these features.

There are some FIXMEs:
* Currently the result is always the default; I've not yet worked on how to resolve this, if it can be;
* I'd like to optionally have the message->response be output, which is useful when debugging tests.

Also, I'm sure there is mypy tags to adjust, but I wanted to get this out to be looked over!

smarx commented:
*Automated message from Dropbox CLA bot*

@neiljp, it looks like you've already signed the Dropbox CLA. Thanks!

neiljp commented:
FYI I plan to investigate changing the expected input to a list of tuples instead of a dict; does that sound good to others?

adnrs96 commented on api/bots_api/bots_test_lib.py:
```diff
@@ -56,42 +56,99 @@ def initialize_bot(self):
+                             if isinstance(i, tuple)
+                             else (dict(default_response_template, content = i), default_method,
+                                   default_result)
+                             for i in resp]
```
We should definitely write this using top down old fashion approach.

neiljp replied:
Right; this line grew somewhat, as I was developing. It is a lot clearer now, though I would love a comprehension syntax that was cleaner.

adnrs96 replied:
I guess it would be really tricky to get a comprehension syntax for logic this big, right (in sense of simplicity), and tricky code is what I would try to avoid.
Although I welcome your thoughts on it and possibly any other variant of the comprehension syntax you believe would be better suited, as it maybe just me who maybe feeling this way regarding this sort of syntax.

neiljp replied:
I'm not sure if there is a better comprehension syntax for this.

The updated code is clearer than the original, I believe.

The only option I can think of at the moment is doing multiple comprehensions and putting the separate 'if' statements as filters, but that wouldn't preserve the order of the original list.

adnrs96 commented on api/bots_api/bots_test_lib.py:
```diff
@@ -56,42 +56,99 @@ def initialize_bot(self):
+                             else (dict(default_response_template, content = i), default_method,
+                                   default_result)
+                             for i in resp]
+                self.assert_bot_response(message, responses)
```
aren't we here asserting a message dict against a response list?

neiljp replied:
We're not asserting message==responses, but rather that given an 'input' of message, that we get 'responses'.

adnrs96 commented:
> FYI I plan to investigate changing the expected input to a list of tuples instead of a dict; does that sound good to others?

Sounds good to me.

neiljp commented on api/bots_api/bots_test_lib.py:
```diff
@@ -56,42 +56,99 @@ def initialize_bot(self):
+                for i in resp:
+                    if isinstance(i, tuple):
+                        li = len(i)
+                        if li != 3 and li != 2:
```
Nice. Done.

neiljp commented:
Updated:
* Expected is now a list(tuple(str, Any)), allowing ordered input (since a list, not a dict, though OrderedDict was an alternative), but also duplicate 'str' entries.
* Switched complex list comprehension to be simpler to read.
* Adjusted to more pythonic "if li not in (2,3)" form.

neiljp commented:
Would the logging.* in check_expected_responses be better as simple raising of Exceptions?

adnrs96 commented:
I guess it would be nice to display error messages. We even use it in number of place in code base. 
Kind of used for generating logs though, which I don't think might be a requirement of a testing code.

neiljp commented:
@adnrs96 I've updated the error handling to raise Exceptions (with improved text, I feel) when the expected content is invalid.

The reporting of different issues in call_request is also much improved, and shows the details of any test failure.

showell commented:
The use of tuples here is pretty brittle and makes the code hard to read.  We should probably discuss on chat whether there's a cleaner approach here.

timabbott commented:
We moved the code in question to the python-zulip-api repository, so I'm closing this PR.  Feel free to reopen a version on the new repository.
